File: docs/dfs/2360_hard/longest_cycle.py

class LongestCycle:  # LeetCode Q.2360.
    def __init__(self):
        self.edges: list[int] = []
        self.visited_orders: list[int] = []

        self.current_order = 1
        self.longest_cycle = -1

    def _dfs_cycle(self, node: int) -> None:
        self.visited_orders[node] = self.current_order
        self.current_order += 1  # Increment for the next node.

        tgt_node = self.edges[node]
        if tgt_node == -1:  # No outgoing edge.
            self.visited_orders[node] = -2  # Current node finalizes DFS.
            return

        if self.visited_orders[tgt_node] == -1:  # Unvisited yet.
            self._dfs_cycle(tgt_node)

        tgt_node_order = self.visited_orders[tgt_node]

        # Target node still under DFS: a cycle just forms.
        if tgt_node_order > 0:
            # Such a cycle starts at target, and ends at current node.
            cycle_len = self.visited_orders[node] + 1 - tgt_node_order
            if cycle_len > self.longest_cycle:
                self.longest_cycle = cycle_len

        self.visited_orders[node] = -2  # Current node finalizes DFS.

    def compute_longest_cycle_len(self, edges: list[int]) -> int:
        self.edges = edges

        # Special marks: -1 = unvisited yet. -2 = has finalized DFS.
        self.visited_orders = [-1] * len(edges)

        self.current_order = 1
        self.longest_cycle = -1

        for node, _ in enumerate(edges):
            if self.visited_orders[node] == -1:  # Unvisited yet.
                self._dfs_cycle(node)

        return self.longest_cycle

File: docs/dfs/2360_hard/longest_cycle.cpp
#include <vector>
using namespace std;

class LongestCycle // LeetCode Q.2360.
{
private:
    vector<int> graph, visitedOrders;
    int currentOrder = 1, maxCycle = -1;

    void dfsCycle(int node) {
        visitedOrders[node] = currentOrder;
        currentOrder++; // Increment for the next node.

        int tgtNode = graph[node];
        if (tgtNode == -1) // No outgoing edge.
        {
            visitedOrders[node] = -2; // Current node finalizes DFS.
            return;
        }

        if (visitedOrders[tgtNode] == -1) // Unvisited yet.
            dfsCycle(tgtNode);

        int tgtNodeOrder = visitedOrders[tgtNode];

        // Target node is also still under DFS: a cycle just forms.
        if (tgtNodeOrder > 0) {
            // Such a cycle starts at target, and ends at current node.
            int cycleLen = visitedOrders[node] + 1 - tgtNodeOrder;
            if (cycleLen > maxCycle)
                maxCycle = cycleLen;
        }

        visitedOrders[node] = -2; // Current node finalizes DFS.
    }

public:
    int computeLongestCycle(vector<int>& edges) {
        graph = edges;

        // Special marks: -1 = unvisited. -2 = has finalized DFS.
        visitedOrders.assign(edges.size(), -1);

        for (int node = 0; node < edges.size(); node++) {
            if (visitedOrders[node] == -1) // Unvisited yet.
                dfsCycle(node);
        }

        return maxCycle;
    }
};